File: src/main/java/kafkastreams/javaexamples/ArticleCountExample.java
package kafkastreams.javaexamples;

import com.fasterxml.jackson.databind.JsonNode;
import kafkastreams.serdes.JsonNodeSerde;
import org.apache.kafka.common.serialization.Serde;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.streams.StreamsBuilder;
import org.apache.kafka.streams.Topology;
import org.apache.kafka.streams.kstream.*;

public class ArticleCountExample extends KafkaStreamsApp {

    public static void main(String[] args) {
        new ArticleCountExample().start("article-count-app");
    }

    public Topology createTopology(StreamsBuilder builder) {
        Serde<String> strings = Serdes.String();
        Serde<Long> longs = Serdes.Long();
        JsonNodeSerde json = new JsonNodeSerde();

        KStream<String, JsonNode> articles = builder.stream("Articles", Consumed.with(strings, json));

        KGroupedStream<String, JsonNode> grouped = articles
                .groupBy((key, value) -> value.path("site").asText(),
                        Serialized.with(strings, json));

        KTable<String, Long> counts = grouped.count();

        counts.toStream().to("ArticleCounts", Produced.with(strings, longs));

        return builder.build();
    }
}

File: src/main/scala/kafkastreams/scalaexamples/ArticleCountExample.scala
package kafkastreams.scalaexamples

import com.fasterxml.jackson.databind.JsonNode
import kafkastreams.scalautils.JacksonDSL._
import kafkastreams.serdes.JsonNodeSerde
import org.apache.kafka.streams.Topology
import org.apache.kafka.streams.scala.ImplicitConversions._
import org.apache.kafka.streams.scala.Serdes.{Long, String}
import org.apache.kafka.streams.scala.StreamsBuilder

object ArticleCountExample extends App {
  new ArticleCountExample().start("article-count-app")
}

class ArticleCountExample extends KafkaStreamsApp {
  def createTopology(builder: StreamsBuilder): Topology = {
    implicit val json = new JsonNodeSerde

    val articles = builder.stream[String, JsonNode]("Articles")

    val articlesPerSite = articles
      .groupBy(extractSite)
      .count()

    articlesPerSite.toStream.to("ArticleCounts")

    builder.build()
  }

  private def extractSite(key: String, article: JsonNode) = article("site").asText
}
